OpenCL/m24000-pure.cl

```diff
@@ -0,0 +1,881 @@
+    const uint4 tmp1 = (uint4) (digest[4], digest[5], digest[6], digest[7]);
+    #endif
+
+    tmps[gid].P[k + 0] = tmp0;
+    tmps[gid].P[k + 1] = tmp1;
+  }
+}
+
+KERNEL_FQ void m24000_loop (KERN_ATTR_TMPS_ESALT (scrypt_tmp_t, bestcrypt_scrypt_t))
+{
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
+
+  GLOBAL_AS uint4 *d_scrypt0_buf = (GLOBAL_AS uint4 *) d_extra0_buf;
+  GLOBAL_AS uint4 *d_scrypt1_buf = (GLOBAL_AS uint4 *) d_extra1_buf;
+  GLOBAL_AS uint4 *d_scrypt2_buf = (GLOBAL_AS uint4 *) d_extra2_buf;
+  GLOBAL_AS uint4 *d_scrypt3_buf = (GLOBAL_AS uint4 *) d_extra3_buf;
+
+  uint4 X[STATE_CNT4];
+  uint4 T[STATE_CNT4];
+
+  #ifdef _unroll
+  #pragma unroll
+  #endif
+  for (int z = 0; z < STATE_CNT4; z++) X[z] = hc_swap32_4 (tmps[gid].P[z]);
+
+  scrypt_smix (X, T, d_scrypt0_buf, d_scrypt1_buf, d_scrypt2_buf, d_scrypt3_buf);
+
+  #ifdef _unroll
+  #pragma unroll
+  #endif
+  for (int z = 0; z < STATE_CNT4; z++) tmps[gid].P[z] = hc_swap32_4 (X[z]);
+
+  #if SCRYPT_P >= 1
+  for (int i = STATE_CNT4; i < SCRYPT_CNT4; i += STATE_CNT4)
+  {
+    for (int z = 0; z < STATE_CNT4; z++) X[z] = hc_swap32_4 (tmps[gid].P[i + z]);
+
+    scrypt_smix (X, T, d_scrypt0_buf, d_scrypt1_buf, d_scrypt2_buf, d_scrypt3_buf);
+
+    for (int z = 0; z < STATE_CNT4; z++) tmps[gid].P[i + z] = hc_swap32_4 (X[z]);
+  }
+  #endif
+}
+
+KERNEL_FQ void m24000_comp (KERN_ATTR_TMPS_ESALT (scrypt_tmp_t, bestcrypt_scrypt_t))
+{
+  /**
+   * base
+   */
+
+  const u64 gid = get_global_id (0);
+  const u64 lid = get_local_id (0);
+  const u64 lsz = get_local_size (0);
+
+  if (gid >= gid_max) return;
+
+  /**
+   * 2nd pbkdf2, creates B
+   */
+
+  u32 w0[4];
+  u32 w1[4];
+  u32 w2[4];
+  u32 w3[4];
+
+  sha256_hmac_ctx_t ctx;
+
+  sha256_hmac_init_global_swap (&ctx, pws[gid].i, pws[gid].pw_len);
+
+  for (u32 l = 0; l < SCRYPT_CNT4; l += 4)
+  {
+    uint4 tmp;
+
+    tmp = tmps[gid].P[l + 0];
+
+    w0[0] = tmp.x;
+    w0[1] = tmp.y;
+    w0[2] = tmp.z;
+    w0[3] = tmp.w;
+
+    tmp = tmps[gid].P[l + 1];
+
+    w1[0] = tmp.x;
+    w1[1] = tmp.y;
+    w1[2] = tmp.z;
+    w1[3] = tmp.w;
+
+    tmp = tmps[gid].P[l + 2];
+
+    w2[0] = tmp.x;
+    w2[1] = tmp.y;
+    w2[2] = tmp.z;
+    w2[3] = tmp.w;
+
+    tmp = tmps[gid].P[l + 3];
+
+    w3[0] = tmp.x;
+    w3[1] = tmp.y;
+    w3[2] = tmp.z;
+    w3[3] = tmp.w;
+
+    sha256_hmac_update_64 (&ctx, w0, w1, w2, w3, 64);
+  }
+
+  w0[0] = 1;
+  w0[1] = 0;
+  w0[2] = 0;
+  w0[3] = 0;
+  w1[0] = 0;
+  w1[1] = 0;
+  w1[2] = 0;
+  w1[3] = 0;
+  w2[0] = 0;
+  w2[1] = 0;
+  w2[2] = 0;
+  w2[3] = 0;
+  w3[0] = 0;
+  w3[1] = 0;
+  w3[2] = 0;
+  w3[3] = 0;
+
+  sha256_hmac_update_64 (&ctx, w0, w1, w2, w3, 4);
+
+  sha256_hmac_final (&ctx);
+
+  char version = esalt_bufs[DIGESTS_OFFSET].version;
+
+  u32 iv[4] = { 0 };
+
+  u32 res[20]; // full would be 24 x u32 (96 bytes)
+
+  if (version == '8')
+  {
+
+        /**
+     * aes shared
+     */
+
+    #ifdef REAL_SHM
```

Comment:
The initialization of the AES constants can't be done in a branch (is drivers limitation). Please move the part the the beginning of the _comp kernel.